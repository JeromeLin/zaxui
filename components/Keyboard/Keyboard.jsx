import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Icon from '../Icon';

const NUMBER_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'];
const PRICE_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'];
const IDCARD_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'x', '0', 'close'];

class Keyboard extends PureComponent {

  onKeyClick = (e, key) => {
    e.preventDefault();
    if (key.length === 0) return;

    const { onKeyClick } = this.props;
    typeof onKeyClick === 'function' && onKeyClick(key);
  }

  getKeys = () => {
    const { type } = this.props;
    switch (type) {
      case 'price':
        return PRICE_KEYS;

      case 'idcard':
        return IDCARD_KEYS;

      default:
        return NUMBER_KEYS;
    }
  }

  render() {
    const { prefixCls } = this.props;

    return (
      <div className={prefixCls}>
        <div className={`${prefixCls}-key`}>
          {
            this.getKeys().map((text, i) => {
              const keyCls = classnames(`${prefixCls}-key-item`, {
                [`${prefixCls}-key-item-disabled`]: text.length === 0,
              });
              return <div className={keyCls} key={+i} onTouchStart={e => this.onKeyClick(e, text)}>{(text === 'close') ? <Icon type="keyboard" /> : text}</div>;
            })
          }
        </div>
        <div className={`${prefixCls}-handle`}>
          <div className={`${prefixCls}-handle-item`} onTouchStart={e => this.onKeyClick(e, 'delete')}><Icon type="deletekey" /></div>
          <div className={`${prefixCls}-handle-item ${prefixCls}-handle-item-ok`} onClick={e => this.onKeyClick(e, 'ok')}>确定</div>
        </div>
      </div>
    );
  }
}

Keyboard.propTypes = {
  prefixCls: PropTypes.string,  // eslint-disable-line
  className: PropTypes.string,  // eslint-disable-line
  type: PropTypes.oneOf(['number', 'price', 'idcard']),
  onKeyClick: PropTypes.func,
};

Keyboard.defaultProps = {
  prefixCls: 'za-keyboard',
  type: 'number',
};

export default Keyboard;
