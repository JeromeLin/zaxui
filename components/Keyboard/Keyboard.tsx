import React, { PureComponent } from 'react';
import classnames from 'classnames';
import PropsType from './PropsType';
import Icon from '../Icon';

const NUMBER_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'];
const PRICE_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'];
const IDCARD_KEYS = ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'x', '0', 'close'];

export interface KeyboardProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class Keyboard extends PureComponent<KeyboardProps, {}> {

  static Picker: any;

  static defaultProps = {
    prefixCls: 'za-keyboard',
    type: 'number',
  };

  onKeyClick = (key) => {
    if (key.length === 0) {
      return;
    }

    const { onKeyClick } = this.props;
    if (typeof onKeyClick === 'function') {
      onKeyClick(key);
    }
  }

  getKeys = () => {
    const { type } = this.props;
    switch (type) {
      case 'price':
        return PRICE_KEYS;

      case 'idcard':
        return IDCARD_KEYS;

      default:
        return NUMBER_KEYS;
    }
  }

  renderKey = (text, index) => {
    const { prefixCls } = this.props;

    const keyCls = classnames(`${prefixCls}-key-item`, {
      [`${prefixCls}-key-item-disabled`]: text.length === 0,
    });

    return (
      <div
        className={keyCls}
        key={+index}
        onClick={() => this.onKeyClick(text)}
      >
        {(text === 'close') ? <Icon type="keyboard" /> : text}
      </div>
    );
  }

  render() {
    const { prefixCls } = this.props;

    return (
      <div className={prefixCls}>
        <div className={`${prefixCls}-key`}>
          {this.getKeys().map(this.renderKey)}
        </div>
        <div className={`${prefixCls}-handle`}>
          <div className={`${prefixCls}-handle-item`} onClick={() => this.onKeyClick('delete')}>
            <Icon type="deletekey" />
          </div>
          <div className={`${prefixCls}-handle-item ${prefixCls}-handle-item-ok`} onClick={() => this.onKeyClick('ok')}>
            确定
          </div>
        </div>
      </div>
    );
  }
}
