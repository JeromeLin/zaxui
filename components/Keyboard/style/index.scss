@import "../../style/core/index";
@import "../../style/variables";
@import "../../Button/style/component";
@import "./component";
