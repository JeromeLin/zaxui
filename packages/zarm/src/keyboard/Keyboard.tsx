import * as React from 'react';
import classnames from 'classnames';
import { Keyboard as KeyboardIcon, DeleteKey as DeleteKeyIcon } from '@zarm-design/icons';
// import useLongPress from '../hooks/useLongPress';
import type PropsType from './PropsType';

type KeyType = Exclude<PropsType['type'], undefined>;

const KEYS: { [type in KeyType]: readonly string[] } = {
  number: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '', '0', 'close'],
  price: ['1', '2', '3', '4', '5', '6', '7', '8', '9', '.', '0', 'close'],
  idcard: ['1', '2', '3', '4', '5', '6', '7', '8', '9', 'X', '0', 'close'],
};

// let longPressTimer: number | null;

export interface KeyboardProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

const Keyboard = React.forwardRef<unknown, KeyboardProps>(
  (
    { prefixCls = 'za-keyboard', type = 'number', onKeyClick, className, locale }: KeyboardProps,
    ref,
  ) => {
    const keyboardRef = (ref as any) || React.createRef<HTMLElement>();

    const cls = classnames(prefixCls, className);
    const getKeys = KEYS[type];

    const onKeyPress = (e, key: string) => {
      e.stopPropagation();

      if (key.length === 0) {
        return;
      }

      if (typeof onKeyClick === 'function') {
        onKeyClick(key);
      }
    };

    // todo: 长按连续删除还有点问题
    // const longPressEvent = useLongPress({},
    //   (e) => {
    //     longPressTimer = window.setInterval(() => {
    //       onKeyPress(e, 'delete');
    //     }, 200);
    //   },
    //   (e) => onKeyPress(e, 'delete'),
    //   () => clearInterval(longPressTimer!),
    // );

    const renderKey = (text: string, index: number) => {
      const keyCls = classnames(`${prefixCls}__item`, {
        [`${prefixCls}__item--disabled`]: text.length === 0,
      });

      return (
        <div className={keyCls} key={+index} onClick={(e) => onKeyPress(e, text)}>
          {text === 'close' ? <KeyboardIcon size="lg" /> : text}
        </div>
      );
    };

    return (
      <div className={cls} ref={keyboardRef}>
        <div className={`${prefixCls}__keys`}>{getKeys.map(renderKey)}</div>
        <div className={`${prefixCls}__handle`}>
          <div
            className={`${prefixCls}__item`}
            onClick={(e) => onKeyPress(e, 'delete')}
            // {...longPressEvent}
          >
            <DeleteKeyIcon size="lg" />
          </div>
          <div
            className={`${prefixCls}__item ${prefixCls}__item--ok`}
            onClick={(e) => onKeyPress(e, 'ok')}
          >
            {locale!.okText}
          </div>
        </div>
      </div>
    );
  },
);

Keyboard.displayName = 'Keyboard';

export default Keyboard;
