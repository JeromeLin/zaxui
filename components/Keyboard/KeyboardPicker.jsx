import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Keyboard from './Keyboard';
import Popup from '../Popup';

class KeyboardPicker extends PureComponent {

  constructor(props) {
    super(props);
    this.state = {
      visible: props.visible || false,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.visible) {
      this.open();
    } else {
      this.close();
    }
  }

  onKeyClick = (key) => {
    if (['ok', 'close'].indexOf(key) > -1) {
      this.close();
      return;
    }
    const { onKeyClick } = this.props;
    typeof onKeyClick === 'function' && onKeyClick(key);
  }

  open = () => {
    this.setState({ visible: true });
  }

  close = () => {
    this.setState({ visible: false });
  }

  render() {
    const { prefixCls, onKeyClick, type, ...others } = this.props;
    return (
      <Popup
        {...others}
        visible={this.state.visible}
        mask={false}
        direction="bottom">
        <Keyboard type={type} onKeyClick={this.onKeyClick} />
      </Popup>
    );
  }
}

KeyboardPicker.propTypes = {
  prefixCls: PropTypes.string,  // eslint-disable-line
  className: PropTypes.string,  // eslint-disable-line
  type: PropTypes.oneOf(['number', 'price', 'idcard']),
  onKeyClick: PropTypes.func,
};

KeyboardPicker.defaultProps = {
  prefixCls: 'za-keyboard',
  type: 'number',
};

export default KeyboardPicker;
