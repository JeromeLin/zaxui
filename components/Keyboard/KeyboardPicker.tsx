import React, { PureComponent } from 'react';
import PropsType from './PropsType';
import Keyboard from './Keyboard';
import Popup from '../Popup';

export interface KeyboardProps extends PropsType {
  prefixCls?: string;
  className?: string;
}

export default class KeyboardPicker extends PureComponent<KeyboardProps, any> {

  static defaultProps = {
    prefixCls: 'za-keyboard',
    type: 'number',
  };

  constructor(props) {
    super(props);
    this.state = {
      visible: props.visible || false,
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.visible) {
      this.open();
    } else {
      this.close();
    }
  }

  onKeyClick = (key) => {
    if (['ok', 'close'].indexOf(key) > -1) {
      this.close();
      return;
    }
    const { onKeyClick } = this.props;
    if (typeof onKeyClick === 'function') {
      onKeyClick(key);
    }
  }

  open = () => {
    this.setState({ visible: true });
  }

  close = () => {
    this.setState({ visible: false });
  }

  render() {
    const { prefixCls, onKeyClick, type, ...others } = this.props;
    return (
      <Popup
        {...others}
        visible={this.state.visible}
        mask={false}
        direction="bottom"
      >
        <Keyboard type={type} onKeyClick={this.onKeyClick} />
      </Popup>
    );
  }
}
