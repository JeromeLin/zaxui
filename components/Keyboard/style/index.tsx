import '../../style/';
import '../../Button/style/';
import '../../Icon/style/';
import './index.scss';
