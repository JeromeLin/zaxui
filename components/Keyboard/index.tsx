import Keyboard from './Keyboard';
import KeyboardPicker from './KeyboardPicker';

Keyboard.Picker = KeyboardPicker;

export default Keyboard;
