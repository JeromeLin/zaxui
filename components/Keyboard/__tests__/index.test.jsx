import React from 'react';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Keyboard from '../index';

describe('Keyboard', () => {
  it('renders correctly', () => {
    const wrapper = render(<Keyboard />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is price', () => {
    const wrapper = render(<Keyboard type="price" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('type is idcard', () => {
    const wrapper = render(<Keyboard type="idcard" />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('click keyboard', () => {
    const onKeyClick = jest.fn();
    const wrapper = shallow(<Keyboard onKeyClick={onKeyClick} />);
    const keys = wrapper.find('.za-keyboard-key');
    expect(keys.childAt(0).text()).toBe('1');
    keys.childAt(0).simulate('click');
    expect(onKeyClick).toBeCalledWith('1');
    keys.childAt(9).simulate('click');
    keys.childAt(11).simulate('click');
    expect(onKeyClick).toBeCalledWith('close');

    const handles = wrapper.find('.za-keyboard-handle');
    handles.childAt(0).simulate('click');
    expect(onKeyClick).toBeCalledWith('delete');
    handles.childAt(1).simulate('click');
    expect(onKeyClick).toBeCalledWith('ok');
  });
});

describe('Keyboard.Picker', () => {
  it('renders correctly', () => {
    const wrapper = render(<Keyboard.Picker />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('visible', () => {
    const wrapper = shallow(<Keyboard.Picker />);
    wrapper.setProps({ visible: true });
    wrapper.setProps({ visible: false });
  });

  it('click keyboard', () => {
    const onKeyClick = jest.fn();
    const wrapper = shallow(<Keyboard.Picker visible onKeyClick={onKeyClick} />);
    const keys = wrapper.find(Keyboard).dive().find('.za-keyboard-key');
    const handles = wrapper.find(Keyboard).dive().find('.za-keyboard-handle');
    keys.childAt(0).simulate('click');
    expect(onKeyClick).toBeCalledWith('1');
    handles.childAt(1).simulate('click');
  });
});
