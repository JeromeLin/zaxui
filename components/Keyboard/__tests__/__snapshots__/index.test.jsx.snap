// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Keyboard renders correctly 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard-key"
  >
    <div
      class="za-keyboard-key-item"
    >
      1
    </div>
    <div
      class="za-keyboard-key-item"
    >
      2
    </div>
    <div
      class="za-keyboard-key-item"
    >
      3
    </div>
    <div
      class="za-keyboard-key-item"
    >
      4
    </div>
    <div
      class="za-keyboard-key-item"
    >
      5
    </div>
    <div
      class="za-keyboard-key-item"
    >
      6
    </div>
    <div
      class="za-keyboard-key-item"
    >
      7
    </div>
    <div
      class="za-keyboard-key-item"
    >
      8
    </div>
    <div
      class="za-keyboard-key-item"
    >
      9
    </div>
    <div
      class="za-keyboard-key-item za-keyboard-key-item-disabled"
    />
    <div
      class="za-keyboard-key-item"
    >
      0
    </div>
    <div
      class="za-keyboard-key-item"
    >
      <i
        class="za-icon za-icon-keyboard"
      />
    </div>
  </div>
  <div
    class="za-keyboard-handle"
  >
    <div
      class="za-keyboard-handle-item"
    >
      <i
        class="za-icon za-icon-deletekey"
      />
    </div>
    <div
      class="za-keyboard-handle-item za-keyboard-handle-item-ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard type is idcard 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard-key"
  >
    <div
      class="za-keyboard-key-item"
    >
      1
    </div>
    <div
      class="za-keyboard-key-item"
    >
      2
    </div>
    <div
      class="za-keyboard-key-item"
    >
      3
    </div>
    <div
      class="za-keyboard-key-item"
    >
      4
    </div>
    <div
      class="za-keyboard-key-item"
    >
      5
    </div>
    <div
      class="za-keyboard-key-item"
    >
      6
    </div>
    <div
      class="za-keyboard-key-item"
    >
      7
    </div>
    <div
      class="za-keyboard-key-item"
    >
      8
    </div>
    <div
      class="za-keyboard-key-item"
    >
      9
    </div>
    <div
      class="za-keyboard-key-item"
    >
      x
    </div>
    <div
      class="za-keyboard-key-item"
    >
      0
    </div>
    <div
      class="za-keyboard-key-item"
    >
      <i
        class="za-icon za-icon-keyboard"
      />
    </div>
  </div>
  <div
    class="za-keyboard-handle"
  >
    <div
      class="za-keyboard-handle-item"
    >
      <i
        class="za-icon za-icon-deletekey"
      />
    </div>
    <div
      class="za-keyboard-handle-item za-keyboard-handle-item-ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard type is price 1`] = `
<div
  class="za-keyboard"
>
  <div
    class="za-keyboard-key"
  >
    <div
      class="za-keyboard-key-item"
    >
      1
    </div>
    <div
      class="za-keyboard-key-item"
    >
      2
    </div>
    <div
      class="za-keyboard-key-item"
    >
      3
    </div>
    <div
      class="za-keyboard-key-item"
    >
      4
    </div>
    <div
      class="za-keyboard-key-item"
    >
      5
    </div>
    <div
      class="za-keyboard-key-item"
    >
      6
    </div>
    <div
      class="za-keyboard-key-item"
    >
      7
    </div>
    <div
      class="za-keyboard-key-item"
    >
      8
    </div>
    <div
      class="za-keyboard-key-item"
    >
      9
    </div>
    <div
      class="za-keyboard-key-item"
    >
      .
    </div>
    <div
      class="za-keyboard-key-item"
    >
      0
    </div>
    <div
      class="za-keyboard-key-item"
    >
      <i
        class="za-icon za-icon-keyboard"
      />
    </div>
  </div>
  <div
    class="za-keyboard-handle"
  >
    <div
      class="za-keyboard-handle-item"
    >
      <i
        class="za-icon za-icon-deletekey"
      />
    </div>
    <div
      class="za-keyboard-handle-item za-keyboard-handle-item-ok"
    >
      确定
    </div>
  </div>
</div>
`;

exports[`Keyboard.Picker renders correctly 1`] = `
<div
  class="za-popup za-popup-hidden"
>
  <div
    class="za-popup-wrapper za-popup-wrapper-bottom"
    style="-webkit-transition-duration:200ms;transition-duration:200ms;"
  >
    <div
      class="za-keyboard"
    >
      <div
        class="za-keyboard-key"
      >
        <div
          class="za-keyboard-key-item"
        >
          1
        </div>
        <div
          class="za-keyboard-key-item"
        >
          2
        </div>
        <div
          class="za-keyboard-key-item"
        >
          3
        </div>
        <div
          class="za-keyboard-key-item"
        >
          4
        </div>
        <div
          class="za-keyboard-key-item"
        >
          5
        </div>
        <div
          class="za-keyboard-key-item"
        >
          6
        </div>
        <div
          class="za-keyboard-key-item"
        >
          7
        </div>
        <div
          class="za-keyboard-key-item"
        >
          8
        </div>
        <div
          class="za-keyboard-key-item"
        >
          9
        </div>
        <div
          class="za-keyboard-key-item za-keyboard-key-item-disabled"
        />
        <div
          class="za-keyboard-key-item"
        >
          0
        </div>
        <div
          class="za-keyboard-key-item"
        >
          <i
            class="za-icon za-icon-keyboard"
          />
        </div>
      </div>
      <div
        class="za-keyboard-handle"
      >
        <div
          class="za-keyboard-handle-item"
        >
          <i
            class="za-icon za-icon-deletekey"
          />
        </div>
        <div
          class="za-keyboard-handle-item za-keyboard-handle-item-ok"
        >
          确定
        </div>
      </div>
    </div>
  </div>
</div>
`;
